File: src/lib/bindings/update.ts
import { invoke } from "@tauri-apps/api/core";

// A newer release the backend verified: round-trips back to install_update.
export interface UpdateInfo {
  version: string;
  url: string;
  signature: string;
}

// `null` means we're already on the latest version.
export const checkUpdate = () => invoke<UpdateInfo | null>("check_update");
export const installUpdate = (info: UpdateInfo) => invoke<void>("install_update", { info });
export const relaunchApp = () => invoke<void>("relaunch_app");
export const openReleasePage = () => invoke<void>("open_release_page");

File: src-tauri/src/commands/update.rs
use crate::services::update::{self, UpdateInfo};

/// Check the release feed. `Some` carries the verified update the frontend then
/// hands back to `install_update`; `None` means we're already current.
#[tauri::command]
pub async fn check_update() -> Result<Option<UpdateInfo>, String> {
    update::check().await
}

/// Download the new executable, verify its signature, and swap it in place.
#[tauri::command]
pub async fn install_update(info: UpdateInfo) -> Result<(), String> {
    update::download_and_apply(&info).await
}

/// Relaunch into the just-installed binary.
#[tauri::command]
pub fn relaunch_app(app: tauri::AppHandle) {
    app.restart();
}

/// Open the releases page - fallback when the in-app update can't be applied.
#[tauri::command]
pub fn open_release_page() {
    update::open_release_page();
}
